File: greedy/swift/135.candy.swift
/*
 * @lc app=leetcode id=135 lang=swift
 *
 * [135] Candy
 */

// @lc code=start
class Solution {
    func candy(_ ratings: [Int]) -> Int {
        var candies = Array(repeating: 1, count: ratings.count)
        for i in 1..<ratings.count {
            if ratings[i] > ratings[i - 1] {
                candies[i] = candies[i - 1] + 1
            }
        }
        for i in stride(from: ratings.count - 2, through: 0, by: -1) {
            if ratings[i] > ratings[i + 1] && candies[i] <= candies[i + 1]{
                candies[i] = candies[i + 1] + 1
            }
        }
        return candies.reduce(0, +)
    }
}
// @lc code=end

File: greedy/cpp/135.candy.cpp
/*
 * @lc app=leetcode id=135 lang=cpp
 *
 * [135] Candy
 */

// @lc code=start
#include <vector>
#include <iostream>
#include <numeric>
using namespace std;

class Solution {
public:
    int candy(vector<int>& ratings) {
        vector<int> candies(ratings.size(), 1);

        for (auto i = 1; i < ratings.size(); ++i) {
            if (ratings[i] > ratings[i-1]) {
                candies[i] = candies[i-1] + 1;
            }
        }

        for (auto i = 0; i < candies.size(); ++i) {
            cout << candies[i];
        }
        cout << endl;

        for (auto i = ratings.size() - 1; i > 0; --i) {
            if (ratings[i] < ratings[i - 1]) {
                candies[i - 1] = max(candies[i - 1], candies[i] + 1);
            }
        }

        for (auto i = 0; i < candies.size(); ++i) {
            cout << candies[i];
        }
        cout << endl;

        return accumulate(candies.begin(), candies.end(), 0);

    }
};
// @lc code=end

